ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicaThread.java

```diff
@@ -84,4 +96,56 @@ protected void logToExternalTable(List<MessageInfo> messageInfoList, RemoteRepli
+   * @param exchangeMetadataResponse Metadata object exchanged between replicas
+   */
+  @Override
+  public void advanceToken(RemoteReplicaInfo remoteReplicaInfo, ExchangeMetadataResponse exchangeMetadataResponse) {
+    StoreFindToken oldToken = (StoreFindToken) remoteReplicaInfo.getToken();
+    // The parent method sets in-memory token
+    super.advanceToken(remoteReplicaInfo, exchangeMetadataResponse);
+    StoreFindToken token = (StoreFindToken) remoteReplicaInfo.getToken();
+    if (token == null) {
+      azureMetrics.absTokenPersistFailureCount.inc();
+      logger.error("Null token for replica {}", remoteReplicaInfo);
+      return;
+    }
+    if (token.equals(oldToken)) {
+      logger.trace("Skipping token upload due to unchanged token, oldToken = {}, newToken = {}", oldToken, token);
+      return;
+    }
+    logger.trace("replica = {}, token = {}", remoteReplicaInfo, token);
+    // Table entity = Table row
+    // =============================================================================================================
+    // | partition-key | row-key | tokenType | logSegment | offset | storeKey | replicatedUntilUTC   | binaryToken |
+    // =============================================================================================================
+    // | partition     | host1   | Journal   | 3_14       | 12     | none     | 2024_Feb_11_14_21_00 | AAASLKJDFX  |
+    // | partition     | host2   | Index     | 2_12       | 32     | AAEWsXZ  | 2024_Jan_02_13_01_00 | AAAEWODSDS  |
+    // =============================================================================================================
+    long lastOpTime = remoteReplicaInfo.getReplicatedUntilUTC();
+    String partitionKey = String.valueOf(remoteReplicaInfo.getReplicaId().getPartitionId().getId());
+    String rowKey = remoteReplicaInfo.getReplicaId().getDataNodeId().getHostname();
+    TableEntity entity = new TableEntity(partitionKey, rowKey)
+        .addProperty(VcrReplicationManager.BACKUP_NODE, dataNodeId.getHostname())
+        .addProperty(VcrReplicationManager.TOKEN_TYPE,
+            token.getType().toString())
+        .addProperty(VcrReplicationManager.LOG_SEGMENT,
+            token.getOffset() == null ? "none" : token.getOffset().getName().toString())
+        .addProperty(VcrReplicationManager.OFFSET,
+            token.getOffset() == null ? "none" : token.getOffset().getOffset())
+        .addProperty(VcrReplicationManager.STORE_KEY,
+            token.getStoreKey() == null ? "none" : token.getStoreKey().getID())
+        .addProperty(VcrReplicationManager.REPLICATED_UNITL_UTC,
+            lastOpTime == Utils.Infinite_Time ? String.valueOf(Utils.Infinite_Time)
+                : VcrReplicationManager.DATE_FORMAT.format(lastOpTime))
+        .addProperty(VcrReplicationManager.BINARY_TOKEN,
+            token.toBytes());
+    // Now persist the token in cloud
+    if (cloudDestination.upsertTableEntity(azureTableNameReplicaTokens, entity)) {
```

Comment:
How frequent do we update the token?

Reply:
The dominant pieces in the repl-loop are fetch missing blobs from peers and upload them to cloud. The token is no more than 300 bytes. I checked IntraColoOneCycleReplicationTime for corp-cluster and there is no bump in the time starting Feb 3 or 4th, which is when I tested this change in corp on all hosts. However, assuming a slight increase in latency, it is a good trade-off for more convenience in debugging and knowing how far the backup is.

As for the frequency, the replicationFetchSizeInBytes is at 1MB in corp. But the token is not updated every 1MB. If we get 1MB of metadata (blobIDs) from peer server, the token is updated by repl code only after all blobs in that 1MB of metadata have been accounted for, which can be much larger than 1Mb. I can increase this config to 4mb to match with other prod clusters. Cost is not a factor based on the metrics (it comes to 8 cents/min without discounts) and neither is disk IO since Azure can handle a lot more IOPs than a disk.  I can get the exact write-rate using this metric ambryReplicaTokenWriteRate but for that this patch has to be merged and deployed through proper channels.